pilot/pkg/config/coredatamodel/discovery.go

```diff
@@ -38,35 +38,104 @@ var (
 	_ model.ServiceDiscovery = &MCPDiscovery{}
 )
 
-// DiscoveryOptions stores the configurable attributes of a Control
-type DiscoveryOptions struct {
-	Env          *model.Environment
-	ClusterID    string
-	DomainSuffix string
-}
-
-// mixerEnabled checks to see if mixer is enabled in the environment
-// so we can set the UID on eds endpoints
-func (o *DiscoveryOptions) mixerEnabled() bool {
-	return o.Env != nil && o.Env.Mesh != nil && (o.Env.Mesh.MixerCheckServer != "" || o.Env.Mesh.MixerReportServer != "")
-}
-
 // MCPDiscovery provides discovery interface for SyntheticServiceEntries
 type MCPDiscovery struct {
 	*SyntheticServiceEntryController
-	*DiscoveryOptions
+	cacheMutex sync.RWMutex
+	// keys [namespace][endpointIP]
+	cacheByEndpointIP map[string]map[string][]*model.ServiceInstance
+	// keys [namespace][service.Hostname]
+	cacheByHostName map[string]map[host.Name][]*model.ServiceInstance
+	cacheInit       bool
 }
 
 // NewMCPDiscovery provides a new instance of Discovery
-func NewMCPDiscovery(controller CoreDataModel, options *DiscoveryOptions) *MCPDiscovery {
-	return &MCPDiscovery{
+func NewMCPDiscovery(controller CoreDataModel) *MCPDiscovery {
+	discovery := &MCPDiscovery{
 		SyntheticServiceEntryController: controller.(*SyntheticServiceEntryController),
-		DiscoveryOptions:                options,
+		cacheByEndpointIP:               make(map[string]map[string][]*model.ServiceInstance),
+		cacheByHostName:                 make(map[string]map[host.Name][]*model.ServiceInstance),
+	}
+	discovery.RegisterEventHandler(schemas.SyntheticServiceEntry.Type, func(config model.Config, event model.Event) {
+		discovery.HandleCacheEvents(config, event)
+	})
+	return discovery
+
+}
+
+func (d *MCPDiscovery) initializeCache() error {
```

Comment:
How about moving this calling to `Run`, so we donot need `cacheInit ` flag

Reply:
I did initially put this in the Run, the issue was Run gets triggered far too early and at that point controller has not received any config yet.

Reply:
This is initialization, and it is incrementally updated by `HandleCacheEvents`. How does this matter?

Reply:
I was under assumption maybe controller would have some config at that point, but sure I will update.